libcaf_test/caf/test/runner.cpp

```diff
@@ -13,23 +13,64 @@
 #include "caf/config_option_adder.hpp"
 #include "caf/config_option_set.hpp"
 #include "caf/detail/log_level.hpp"
+#include "caf/detail/set_thread_name.hpp"
 
+#include <condition_variable>
 #include <iostream>
+#include <mutex>
 #include <optional>
 #include <regex>
 #include <string>
+#include <thread>
 
 namespace caf::test {
 
 namespace {
 
+class watchdog {
+public:
+  watchdog(int secs) {
+    using detail::format_to;
+    thread_ = secs <= 0 ? std::thread() : std::thread{[=] {
+      caf::detail::set_thread_name("test.watchdog");
+      auto tp = std::chrono::high_resolution_clock::now()
+                + std::chrono::seconds(secs);
+      std::unique_lock<std::mutex> guard{mtx_};
+      while (!canceled_
+             && cv_.wait_until(guard, tp) != std::cv_status::timeout) {
+      }
+      if (!canceled_) {
+        auto err_out = [] { return std::ostream_iterator<char>{std::cerr}; };
+        format_to(err_out(),
+                  "WATCHDOG: unit test did not finish within {} seconds\n",
+                  secs);
+        abort();
+      }
+    }};
+  }
+  ~watchdog() {
+    {
+      std::lock_guard<std::mutex> guard{mtx_};
+      canceled_ = true;
+      cv_.notify_all();
+    }
+    if (thread_.joinable())
+      thread_.join();
```

Comment:
If no thread is running, we don't need to do any locking at all:

```suggestion
    if (!thread_.joinable())
      return;
    {
      std::lock_guard<std::mutex> guard{mtx_};
      canceled_ = true;
      cv_.notify_all();
    }
    thread_.join();
```